### DISCUSSION TITLE
Inconsistent behavior between IDEs when stopping and restarting a program

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13728] [General]
I have a program which uses Bluetooth on the PICO-W. I have two IDEs,  Thonny and VS Code using the Micropico extension. When I power cycle the PICO-W and start the application in either Thonny or VS Code,  I can see the Bluetooth stack (port uses btstack) initializes and that IRKs and other keys are properly generated (for any eventual bondings). I can bond with devices (pairing and bonding) and reconnect (encrypt) properly. If I stop the application in VS Code and then restart it (no power cycle) I can continue to to pair with new devices and properly reconnect to those devices I have paired with. However, on Thonny if I stop the application and restart it, I can no longer pair/bond or reconnect. (No problem with unpairable devices however.) I can see in the Thonny case that it tries to re-initialize bluetooth, but it fails to properly setup the pairing and IRK keys.

I am confused as to this behavior. I also have a similar project which after a power cycle in BOTH Thonny and VS Code, the Bluetooth initializes in the same manner as the Thonny application restart and pairing/bonding does not work.

Is the problem incorrect initialization of the Bluetooth? I do the following at the start of the Bluetooth task:
```
    async def run_scanner(self) -> None:
        global log_level
        global state
        # initialize Bluetooth
        print('LNI: Initialize Bluetooth')
        ble: BLE = BLE()
        ble.active(True)
        ble.config(bond=True)
        ble.config(io=Btle.IO_CAPABILITY_KEYBOARD_DISPLAY)
        ble.config(mitm=True)
        ble.config(le_secure=True)
        # set the log level in the lni_aioble library
        log_level = 2
        state: str = 'running'
        while True:  # Here is where the Bluetooth runs
```
Does this have to be done at a specific time during app spinup? ( **I am using the aioble library as source***.) Is there a limit to the nesting of directories? (I am more concerned about understanding the case where the Bluetooth does not come up correctly - pairing doesnt work but everything else does so with non pairable devices no problem.)

Am I missing something in my shutdown of the task(s) upon program exit? I have two approaches, one where I signal the task to cleanup itself before terminating and another one where I do a simple `bt_task.cancel()`. It makes no difference in the behavior.

I cannot find much information about the details of Bluetooth startup and shutdown and how that relates to restarting a program without a power cycle. Perhaps Micropython programs are not supposed to be restarted without a power cycle?

Any insight would be greatly appreciated!

* I had to use aioble as source as I needed a way to tell aioble that a device issued a security request and bonding/encryption happened in the btstack. I added a method to do that. Otherwise, aioble's DeviceConnection thinks it is not bonded and not encrypted.

**Comment by andrewleech** [comment]
I don't use Thonny so I'm not sure if the issue, but I do know it does run a bunch of "custom" commands on the board to set up its integration.

One of those commands _might_ be changing the "current directory" on the device, which would then change where aioble stores / loads the bond data.

Before you start BLE at all, you can run:
```
import aioble.security
aioble.security.load_secrets(path="/flash/ble_secrets.json")
```
Then continue to run all your other code the same.
Running `load_secrets` manually beforehand let's you set the file to an absolute path that suits your board, rather than the default that stores in the current working directory.

**Reply by brianreinhold** [reply]
I gave it a try and I see the same condition. In fact I don't have to go between two IDEs. If in VS Code I power cycle the PICO-W and then start application, it initializes correctly and the client default security keys (IRKs and LTKs) are properly created by btstack. Bonding and bonded reconnects happen correctly. If I shutdown the app exit VS Code and restart  VS Code and start the app (no power cycle) the btstack security info is not created and no bonding or bonded reconnects function. Unpairable devices work fine.

**Reply by Josverl** [reply]
@brianreinhold  IIRC you use MicroPico - which maintain a separate connection to a Pico - 
and may have some initialization and side effects that are not that apparent. 
perhaps @PaulOber can comment ?

**Reply by brianreinhold** [reply]
IN addition if I hit the refresh button and restart the app I also lose pairing. All else (as far as I can tell) is okay.
